### PR TITLE
`ItemTag.firework_map` + mech support

### PR DESCRIPTION
**Pull request by tal5**
## Additions

- `ItemTag.firework_map` returns the item's firework property value as a MapTag of ordered integers for each effect - `map@[1 = map@[trail = true; flicker = false; type = BALL; color = co@red; fadeColor = co@white]; 2 = map@[trail = false; flicker = true; type = STAR; color = co@green; fadeColor = co@green]]`
- `ItemTag.firework` mechanism now accepts MapTag input in the above format, and clears all firework effects when no input is given
- Update `ItemFirework` property to new tag registration (`PropertyParser#registerTag` instead of `getObjectAttribute`)

### CODE REVIEW

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemFirework.java [hunk 1]
```diff
@@ -75,12 +74,42 @@ public ListTag getFireworkData() {
...
+                }
+                Color ColOne = effect.getColors() != null && effect.getColors().size() > 0 ? effect.getColors().get(0) : Color.BLUE;
+                Color ColTwo = effect.getFadeColors() != null && effect.getFadeColors().size() > 0 ? effect.getFadeColors().get(0) : ColOne;
+                MapTag effectMap = new MapTag();
+                effectMap.putObject("trail", new ElementTag(effect.hasTrail()));
+                effectMap.putObject("flicker", new ElementTag(effect.hasFlicker()));
+                effectMap.putObject("type", new ElementTag(effect.getType().name()));
+                effectMap.putObject("color", new ColorTag(ColOne));
+                effectMap.putObject("fadeColor", new ColorTag(ColTwo));
+                map.putObject(String.valueOf(i+1), effectMap);
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemFirework.java]: This is a list. You've made a list using MapTags. Use a ListTag.

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemFirework.java [hunk 2]
```diff
@@ -75,12 +74,42 @@ public ListTag getFireworkData() {
         return list;
     }
 
-    @Override
-    public ObjectTag getObjectAttribute(Attribute attribute) {
-
-        if (attribute == null) {
-            return null;
+    public MapTag getFireworkDataMap(){
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemFirework.java]: basic java formatting

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemFirework.java [hunk 3]
```diff
@@ -75,12 +74,42 @@ public ListTag getFireworkData() {
...
+            if (power != 0) {
+                map.putObject("power", new ElementTag(power));
+            }
         }
+        else {
+            effects = Collections.singletonList(((FireworkEffectMeta) item.getItemMeta()).getEffect());
+        }
+        if (effects != null) {
+            int size = effects.size();
+            for (int i = 0; i < size; i++) {
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemFirework.java]: This should also be reverted to the way it's done in `getFireworkData`

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemFirework.java [hunk 4]
```diff
@@ -75,12 +74,42 @@ public ListTag getFireworkData() {
...
+                    continue;
+                }
+                Color ColOne = effect.getColors() != null && effect.getColors().size() > 0 ? effect.getColors().get(0) : Color.BLUE;
+                Color ColTwo = effect.getFadeColors() != null && effect.getFadeColors().size() > 0 ? effect.getFadeColors().get(0) : ColOne;
+                MapTag effectMap = new MapTag();
+                effectMap.putObject("trail", new ElementTag(effect.hasTrail()));
+                effectMap.putObject("flicker", new ElementTag(effect.hasFlicker()));
+                effectMap.putObject("type", new ElementTag(effect.getType().name()));
+                effectMap.putObject("color", new ColorTag(ColOne));
+                effectMap.putObject("fadeColor", new ColorTag(ColTwo));
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemFirework.java]: `fadeColor` is not a valid MapTag key

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemFirework.java (hunk 1 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemFirework.java]: Just thought that format would be easier to use - I.e.
```yaml
  - definemap Firework:
      power: 3
      1:
        color: red
        fadeColor: white
        trail: true
      2:
        color: green
        type: star
        flicker: true
```

Comment (mcmonkey4eva) [reply plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemFirework.java]: A tag is not a script. A tag is a tag.

**Comment by tal5** [comment]
Fixes from reviews implemented

File: plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemFirework.java [hunk 5]
```diff
@@ -112,60 +161,141 @@ public String getPropertyId() {
...
+
         // <--[mechanism]
         // @object ItemTag
         // @name firework
         // @input ListTag
         // @description
         // Sets the firework's settings.
-        // Each item in the list is formatted as: TRAIL,FLICKER,TYPE,RED,GREEN,BLUE,RED,GREEN,BLUE
+        // Each item in the list is formatted as:
+        // A list of comma-separated values, in TRAIL,FLICKER,TYPE,RED,GREEN,BLUE,RED,GREEN,BLUE format.
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/properties/item/ItemFirework.java:191]: wat.

**Pull request merged**